core/templates/pages/exploration-player-page/services/conversation-flow.service.spec.ts

```diff
@@ -0,0 +1,103 @@
+// limitations under the License.
+
+/**
+ * @fileoverview Unit tests for conversation flow service.
+ */
+
+import {HttpClientTestingModule} from '@angular/common/http/testing';
+import {NO_ERRORS_SCHEMA} from '@angular/core';
+import {TestBed, waitForAsync} from '@angular/core/testing';
+
+import {ConversationFlowService} from './conversation-flow.service';
+import {StateCard} from 'domain/state_card/state-card.model';
+import {ContentTranslationLanguageService} from './content-translation-language.service';
+import {ContentTranslationManagerService} from './content-translation-manager.service';
+import {ExplorationPlayerStateService} from './exploration-player-state.service';
+import {PlayerTranscriptService} from './player-transcript.service';
+import {TranslateService} from '@ngx-translate/core';
+import {MockTranslateService} from 'components/forms/schema-based-editors/integration-tests/schema-based-editors.integration.spec';
+import {Interaction} from 'domain/exploration/InteractionObjectFactory';
+
+describe('Conversation flow service', () => {
+  let contentTranslationLanguageService: ContentTranslationLanguageService;
+  let contentTranslationManagerService: ContentTranslationManagerService;
+  let conversationFlowService: ConversationFlowService;
+  let explorationPlayerStateService: ExplorationPlayerStateService;
+  let playerTranscriptService: PlayerTranscriptService;
+
+  let displayedCard = new StateCard(
+    null,
+    null,
+    null,
+    new Interaction([], [], null, null, [], '', null),
+    [],
+    null,
+    '',
+    null
+  );
+
+  beforeEach(waitForAsync(() => {
+    TestBed.configureTestingModule({
+      imports: [HttpClientTestingModule],
+      providers: [
+        ConversationFlowService,
+        {
+          provide: TranslateService,
+          useClass: MockTranslateService,
+        },
+      ],
+      schemas: [NO_ERRORS_SCHEMA],
+    });
+
+    contentTranslationLanguageService = TestBed.inject(
+      ContentTranslationLanguageService
+    );
+    contentTranslationManagerService = TestBed.inject(
+      ContentTranslationManagerService
+    );
+    conversationFlowService = TestBed.inject(ConversationFlowService);
+    explorationPlayerStateService = TestBed.inject(
+      ExplorationPlayerStateService
+    );
+    playerTranscriptService = TestBed.inject(PlayerTranscriptService);
+  }));
+
+  it('should handle new card addition', () => {
+    spyOn(playerTranscriptService, 'addNewCard');
+    spyOn(explorationPlayerStateService, 'getLanguageCode').and.returnValue(
+      'en'
+    );
+    spyOn(
+      contentTranslationLanguageService,
+      'getCurrentContentLanguageCode'
+    ).and.returnValue('es');
+    spyOn(
+      contentTranslationManagerService,
+      'displayTranslations'
+    ).and.returnValue();
+
+    conversationFlowService.addNewCardAndDisplayTranslations(displayedCard);
+    expect(playerTranscriptService.addNewCard).toHaveBeenCalled();
+    expect(
```

Comment:
Try to test things that are more than just "this function was called". 

E.g. if language is English, verify that the end result of the function is that the new card shows up in English. And if it's not, verify that the end result is that the new card shows up in the other language. (Also have appropriate pre-checks that run before the 'add new card' function is executed and that verify that those same expectations result in the text from the already-existing card.)

Reply:
Hm, OK, I think it's fine as you have it then, thanks.